contracts/MarginCalculator.sol

```diff
@@ -2,11 +2,280 @@
+
+        // For the currenct version, ensure denominated == short.collateral
+        address shortCollateral = OtokenInterface(_vault.shortOtokens[0]).collateralAsset();
+        require(shortCollateral == _demonimated, "MarginCalculator: Denomintated token should be short.collateral");
+
+        int256 netOtoken = _calculateOtokenNetValue(_vault);
+        int256 totalValue = collateralAmount.add(netOtoken);
+        isExcess = totalValue >= 0;
+        netValue = totalValue.intToUint();
+    }
+
+    /**
+     * @dev Calculate the net value of long token + short token in a vault.
+     * @return netOtoken net worth of long otoken and short otoken
+     */
+    function _calculateOtokenNetValue(Vault memory _vault) internal view returns (int256 netOtoken) {
+        address short = _vault.shortOtokens[0];
+        int256 shortAmount = _vault.shortAmounts.length > 0 ? FixedPointInt256.uintToInt(_vault.shortAmounts[0]) : 0;
+        int256 longAmount = _vault.longAmounts.length > 0 ? FixedPointInt256.uintToInt(_vault.longAmounts[0]) : 0;
+
+        bool expired = now > OtokenInterface(short).expiryTimestamp();
+        bool isPut = OtokenInterface(short).isPut();
+
+        if (!expired) {
+            int256 shortStrike = FixedPointInt256.uintToInt(OtokenInterface(short).strikePrice());
+            int256 longStrike = 0;
+            if (_vault.longOtokens.length > 0)
+                longStrike = FixedPointInt256.uintToInt(OtokenInterface(_vault.longOtokens[0]).strikePrice());
+
+            if (isPut) {
+                netOtoken = _calculateNonExpiredPutNetValue(shortAmount, longAmount, shortStrike, longStrike);
+            } else {
+                netOtoken = _calculateNonExpiredCallNetValue(shortAmount, longAmount, shortStrike, longStrike);
+            }
+        } else {
+            int256 shortCashValue = FixedPointInt256.uintToInt(getExpiredCashValue(short));
+            int256 longCashValue = 0;
+            if (_vault.longOtokens.length > 0)
+                longCashValue = FixedPointInt256.uintToInt(getExpiredCashValue(_vault.longOtokens[0]));
+            if (isPut) {
+                netOtoken = _calculateExpiredPutNetValue(shortAmount, longAmount, shortCashValue, longCashValue);
+            } else {
+                (uint256 underlyingPrice, ) = _getUnderlyingPrice(address(short));
+                int256 underlyingPriceInt = FixedPointInt256.uintToInt(underlyingPrice);
+                netOtoken = _calculateExpiredCallNetValue(
+                    shortAmount,
+                    longAmount,
+                    shortCashValue,
+                    longCashValue,
+                    underlyingPriceInt
+                );
+            }
+        }
+    }
+
+    /**
+     * @dev calculate non-expired net put spread value.
+     *
+     * Formula: net = (long strike * min (short amount, long amount)) - (short amount * short strike)
+     *
+     * @return net value
+     */
+    function _calculateNonExpiredPutNetValue(
```

Comment:
maybe rename to something about difference in strikes or something? or required collat for put spread? @aleone

Reply:
also does it make a difference if expired or not?

Reply:
I does not, the computation is similar `( long cash value * long Amount ) - (short cash value * short amount)`, but when we were writing the doc, we decided not to use a single `getCashValue` function for expired and non-expired function.

That's why we ends up having `getExpiredCashValue` for expired otoken, and use `Otoken.strikePrice()` and not use the term `cashValue` for the non-expired ones.